Adjust memory reconfiguration for oVirt

jhernand opened the pull request:
Changing the memory assigned to virtual machine in oVirt must follow the
following rules:

1. If the virtual machine is running, then the difference between the old
amount of memory and the new amount of memory needs to be a multiple of
256 MiB. The reason for this is that memory is added plugging new
virtual DIMMS to the virtual machine, and the size of these is 256 MiB.

2. The physical memory reserved for the virtual machine should always be
less or equal to the virtual memory.

This patch changes the oVirt provider so that implements these rules,
adjusting the amount requested by the user accordingly. For example, if
the virtual machine is running with 100 MiB of virtual memory, and the
user requests to increase it to 300 MiB, then the provider will adjust
this to 356 MiB, so the delta will be a multiple of 256 MiB.

https://bugzilla.redhat.com/1356468
https://bugzilla.redhat.com/1356475

jhernand commented:
Note that this requires a new version of the `ovirt` gem, including the changes in ManageIQ/ovirt/pull/72.

jhernand commented:
@jelkosz please review.

jhernand commented:
@borod108, @masayag please review.

jhernand commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Bugs 1356468 & 1356475, problems when hot plugging memory not multiple of 256 MiB](https://trello.com/c/zZh0xLQE/87-bugs-1356468-1356475-problems-when-hot-plugging-memory-not-multiple-of-256-mib)

jelkosz commented:
LGTM :+1:

masayag commented on app/models/manageiq/providers/redhat/infra_manager.rb:
```diff
@@ -115,16 +122,51 @@ def prepare_disk(disk_spec, ems_storage_uid)
+        memory = current_memory + delta
+        _log.info(
+          "The change in virtual memory needs to be a multiple of #{HOT_PLUG_DIMM_SIZE / 1.megabyte} MiB, " +
+          "so it will be adjusted to #{memory / 1.megabyte} MiB."
```
I'd include also the vm name to provide a VM context for this message.

masayag commented on app/models/manageiq/providers/redhat/infra_manager.rb:
```diff
@@ -115,16 +122,51 @@ def prepare_disk(disk_spec, ems_storage_uid)
+          "so it will be adjusted to #{memory / 1.megabyte} MiB."
+        )
+      end
+    end
```
lines 132-148 can be extracted to a method which is responsible only for calculating the memory.

masayag commented on app/models/manageiq/providers/redhat/infra_manager.rb:
```diff
@@ -115,16 +122,51 @@ def prepare_disk(disk_spec, ems_storage_uid)
+      guaranteed = memory
+      _log.info(
+        "The guaranteed physical memory needs to be less or equal than the virtual memory, so it will be " +
+        "adjusted to #{guaranteed / 1.megabyte} MiB."
```
same comment about including the vm name/id as part of the message

masayag commented on app/models/manageiq/providers/redhat/infra_manager.rb:
```diff
@@ -115,16 +122,51 @@ def prepare_disk(disk_spec, ems_storage_uid)
+        "The guaranteed physical memory needs to be less or equal than the virtual memory, so it will be " +
+        "adjusted to #{guaranteed / 1.megabyte} MiB."
+      )
+    end
```
lines 150-161 can be extracted to a method which is responsible for calculating the guaranteed memory

jhernand commented:
Rebased and adjusted to avoid changing the Gemfile, as that is now part of the `manageiq-gems-pending` project, see ManageIQ/manageiq-gems-pending/pull/7.

jhernand commented:
Thanks @masayag, I am working on your suggestions.

jhernand commented:
Updated to adress the comments from @masayag and to add some additional test cases.

jhernand commented:
Addressed Rubocop complaints.

jhernand commented:
@durandom please review.

durandom commented:
@jhernand if you think this is good to go from your team perspective and members of your team reviewed this (as they apparently did) you can ask miq-bot to assign the PR to me, then I'll do the final review. Thanks

@miq-bot assign @durandom

durandom commented:
@oourfali can you apply the backport labels?

jhernand commented:
@agrare, please note that before merging this we need to merge ManageIQ/ovirt/pull/72. Then we need to release the new version of the `ovirt` gem. Then we need to update `manageiq-gems-pending`, ManageIQ/manageiq-gems-pending/pull/7, and then, finally, this can be merged.

agrare commented:
Thanks @jhernand ping me when this is ready to go

jhernand commented:
@agrare version 0.14.0 of the `ovirt` gem has been released:

  https://rubygems.org/gems/ovirt/versions/0.14.0

And the `manageiq-gems-pending` project has been updated to use it. Can we now merge this pull request?

jhernand commented:
Thanks @agrare .

As the [bug](https://bugzilla.redhat.com/1356468) that this fixes is targeted for CFME 5.7.1, I am adding the backport label.

@miq-bot add_label euwe/yes

@bdunne, @agrade, how do we handle the backport in the other gemes that are needed, `ovirt` and `manageiq-gems-pending`?

bdunne commented:
@jhernand You can add the `euwe/yes` label to the `manageiq-gems-pending` PR and it will be backported for you.

jhernand commented:
OK, thanks @bdunne, doing it.

simaishi commented:
Euwe backport details:

```
$ git log -1
commit 06f9617f4e61a231d2ce1a003b203ef3391739d6
Author: Adam Grare <agrare@redhat.com>
Date:   Tue Nov 22 07:45:30 2016 -0500

    Merge pull request #12550 from jhernand/adjust_memory_to_satisfy_ovirt_constraints
    
    Adjust memory reconfiguration for oVirt
    (cherry picked from commit 4994ec87256403c1a21d924abc4b8bd1cb406360)
    
    https://bugzilla.redhat.com/show_bug.cgi?id=1404316
```
